sockeye/lexical_constraints.py

```diff
@@ -24,106 +24,139 @@
 # Represents a list of raw constraints for a sentence. Each constraint is a list of target-word IDs.
 RawConstraintList = List[List[int]]
 
+class AvoidTrie:
```

Comment:
Hmm, it is a bit of a strange Trie. The nodes at level i contains a set of phrases the of length i+1 (with root being level 0), instead of the implicit representation of the path up to the node. Is this because of some optimization?

 **Update:** Probably this is to support the avoid() function below, right?

Reply:
Yes, to support `avoid()`. At each trie node, we record all words that should not be generated from that state. A more principled trie would generate arcs leading to terminal states, but this is more efficient for our purposes. Arcs are only generated if they lead to internal states of a phrase that must be avoided.